### DISCUSSION TITLE
Nginx Container Becomes Unresponsive After Esmerald Deployment: Debugging SWAG + Esmerald Deployment Issue

### DISCUSSION THREAD

**Discussion by sanket-parte** [dymmond/esmerald#375] [General]
Hi Everyone,

We have deployed the Esmerald application using Docker, and it's being served through nginx configured in the [linuxserver/swag](https://hub.docker.com/r/linuxserver/swag) docker container. We’re encountering an issue where, after starting the Esmerald application, the server becomes unresponsive after some time. Specifically, the swag (nginx) container stops responding, leading to request timeout errors when accessing APIs.

Here are the current configurations:
**Uvicorn Script**
We are using the following uvicorn script to start the Esmerald application:
```
import uvicorn
from esmerald.conf import settings

if __name__ == "__main__":
    """
    Uses the `ESMERALD_SETTINGS_MODULE` settings configuration and loads \
    accordingly.
    """
    uvicorn.run(
        "main:app",
        port=8000,
        host="0.0.0.0",
        reload=True,
        lifespan="on",
        log_level="debug",
    )
```

**Nginx Configuration**
Here is the nginx configuration:
```
location /project/api/ {
    proxy_set_header Host $http_host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;
    proxy_set_header X-CSS-Protection "1; mode=block";
    proxy_set_header X-Content-Type-Options "nosniff";
    proxy_set_header Cache-Control "public,max-age=120,must-revalidate,s-maxage=120";
    add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
    add_header Content-Security-Policy "default-src 'self';" always;
    add_header X-Frame-Options "SAMEORIGIN";
    proxy_redirect off;
    proxy_buffering off;
    proxy_pass http://project_service:8000/project/api/;
}
```

Can anyone suggest how to debug this issue?

**Comment by tarsil** [comment]
@sanket-parte It seems like you might have some resources issues or some issues with configurations. You example is not enough to understand the cause but have a look at this 
https://github.com/dymmond/esmerald-tortoise-nginx-gunicorn-supervisor

It's public archive but there also has some nginx, Gunicorn, docker and uvicorn configurations that you can probably check and use?

**Comment by tarsil** [comment]
Ohhhh wait. @sanket-parte that script to start uvicorn it won't work properly in n production. They even said that you should use that for development only (https://github.com/dymmond/esmerald/blob/main/esmerald/conf/directives/project_template/project_name/serve.py-tpl#L5).

The URL I sent before has an example of configurations how to start a server for production using also supervisor. 

https://github.com/dymmond/esmerald-tortoise-nginx-gunicorn-supervisor

Have a look and see how we did.

Also in the docs of Esmerald we have an example as well

**Comment by sanket-parte** [comment]
Hey @tarsil, Thank you very much! for your quick response. I will try this configurations.

**Comment by sanket-parte** [comment]
Hey @tarsil , 

I’ve tried configuring Gunicorn using the setup provided in [this URL](https://github.com/dymmond/esmerald-tortoise-nginx-gunicorn-supervisor) that you shared, but the same issue persists.

Here are the configurations we're using:

### Gunicorn Configuration Script::
```python
import json
import multiprocessing
import os

proc_name = os.getenv("GUNIICORN_PROCESS_NAME", "gunicorn")
project_dir = os.getenv("PROJECT_DIR", "/app")
chdir = os.path.join(project_dir, "/project")
pidfile = os.getenv("GUNIICORN_PIDFILE", None)
host = os.getenv("GUNIICORN_HOST", "0.0.0.0")
port = os.getenv("GUNIICORN_PORT", "8000")
keepalive = int(os.getenv("GUNIICORN_KEEP_ALIVE", "5"))
timeout = int(os.getenv("GUNIICORN_TIMEOUT", "120"))
graceful_timeout = int(os.getenv("GUNIICORN_GRACEFUL_TIMEOUT", "120"))
forwarded_allow_ips = os.getenv("GUNIICORN_FORWARD_ALLOW_IPS", "*")
worker_class = os.getenv("GUNIICORN_WORKER_CLASS", "uvicorn.workers.UvicornWorker")
user = os.getenv("USER", "project_user")
group = os.getenv("GROUP", "project_user")
workers_per_core = float(os.getenv("GUNIICORN_WORKERS_PER_CORE", "1"))

bind = os.getenv("BIND", f"{host}:{port}")

# Logs configurations
loglevel = os.getenv("GUNIICORN_LOGLEVEL", "info")
logger_path = os.getenv("LOGS_DIR")
accesslog = "-"
errorlog = os.path.join(logger_path, "gunicorn-error.log")

web_concurrency = None
max_workers = os.getenv("GUNIICORN_MAX_WORKERS")
use_max_workers = None

if max_workers:
    use_max_workers = int(max_workers)

concurr = os.getenv("GUNIICORN_WEB_CONCURRENCY")
cores = multiprocessing.cpu_count()
default_web_concurrency = float(workers_per_core) * cores

if concurr:
    web_concurrency = int(concurr)
    assert web_concurrency > 0
else:
    web_concurrency = max(int(default_web_concurrency), 2)
    if use_max_workers:
        web_concurrency = min(web_concurrency, use_max_workers)

```

### **Supervisord Configuration FIle**: 
```ini
[program:gunicorn]
command = gunicorn main:app -c /app/conf/gunicorn-conf.py
user = project_user
numprocs = 1
priority = 14
startsecs = 10
autostart = true
autorestart = true
process_name = gunicorn-%(process_num)d
stdout_logfile = /dev/stdout
stdout_logfile_maxbytes = 0
redirect_stderr = true
```

### Command to Start Supervisord:
```sh
/usr/bin/supervisord
```

Could you please assist us in debugging this issue?

**Reply by tarsil** [reply]
@sanket-parte first things first.

1. You call your env vars GUNIICORN and not GUNICORN, is that intentional?
2. In your docker did you place the config inside the "/app/conf" and called it gunivorn-conf.py?
3. Your user used in the supervisor has enough permissions? 
4. Did you copy the supervisor.conf to the proper location to be used? 


Show me the log from when you start the container to see the errors, please

**Comment by sanket-parte** [comment]
Yes, all these configurations are correct. I am able to run the application, and it is accessible from the browser. However, after some time, my Nginx container becomes unresponsive, leading to a gateway timeout error in the browser.

Following are the logs after running the application using above configurations:

![gunicorn_logs](https://github.com/user-attachments/assets/da9e65ef-0b58-4bc0-a2d7-c610db8cd8f2)

Access logs:
![gunicorn_access_logs](https://github.com/user-attachments/assets/b1b88389-970f-4442-9648-cb0595068570)

**Comment by tarsil** [comment]
This is very odd. Do you use any "uvicorn.run()"? You should not do that

**Reply by tarsil** [reply]
I'm not sure if you also have enough resources for docker to run. This can also be an issue

**Reply by sanket-parte** [reply]
No, now we are not using uvicorn.run()

Below is the main.py file, which is run by Gunicorn to start the application:

```python
import os
import sys
from esmerald import Esmerald, Include, Gateway, Inject, OpenAPIConfig
from esmerald.exceptions import HTTPException, ValidationErrorException
from esmerald.contrib.schedulers.asyncz.config import AsynczConfig
from asyncz.stores.mongo import MongoDBStore
from pymongo import MongoClient
from esmerald.conf import settings


def build_path():
    """
    Builds the path of the project and project root.
    """
    SITE_ROOT = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
    if SITE_ROOT not in sys.path:
        sys.path.append(SITE_ROOT)

def get_scheduler_config() -> AsynczConfig:
    """
    Builds the scheduler configuration.
    """
    return AsynczConfig(
        stores={"default": MongoDBStore(client=MongoClient(
            settings.mongodb_database_url,
            connect=False,
            authSource="admin"))},
        tasks={
            "file_processing_scheduler": "project.apps.file_proc.tasks",
            },
    )


def get_application():
    """
    This is optional. The function is only used for organisation purposes.
    """
    build_path()

    # from project.apps.authz.middleware import TokenValidateMiddleware
    from project.apps.authz.v1.views import userlogin
    from project.data_ingestion import loader
    from project.apps.common.custom_interceptors import \
        RequestParamInterceptor
    from project.apps.common.dependencies import database_injection, \
        get_api_logger, redis_db_injection
    from project.apps.common.custom_exceptions import (
        custom_http_exception_handler
        )
    from project.apps.common.v1.views import languages
    from esmerald.conf import settings
    from project.startup_scripts import load_permissions, \
        verify_access_token_version

    openapi_config = OpenAPIConfig(
    title="Project",
    docs_url="/docs/swagger",
    swagger_js_url = 'https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.9.0/swagger-ui-bundle.min.js',
    swagger_css_url = 'https://cdn.jsdelivr.net/npm/swagger-ui-dist@5.9.0/swagger-ui.min.css',
    swagger_favicon_url = 'https://esmerald.dev/statics/images/favicon.ico'
    )

    app = Esmerald(
        routes=[
            Gateway(path=settings.base_url_prefix, handler=languages),
            Gateway(path=settings.base_url_prefix+"/{company_code}/login", handler=userlogin),

            Include(
                path=settings.base_url_prefix,
                namespace="project.urls",
                middleware=[]

            ),
        ],
        on_startup=[load_permissions, verify_access_token_version],
        interceptors=[RequestParamInterceptor],
        dependencies={
            "company_db": Inject(database_injection),
            "redis_db": Inject(redis_db_injection),
            "logger": Inject(get_api_logger)
        },
        exception_handlers={
                        HTTPException: custom_http_exception_handler,
                        ValidationErrorException: custom_http_exception_handler
                        },
        openapi_config=openapi_config
                )
    return app


app = get_application()

```

**Reply by sanket-parte** [reply]
Regarding resources, we are having 12 vCPU plus 48GB ram in the system

**Reply by tarsil** [reply]
This is very odd. This looks like a normal Esmerald application and it should work. As far as Esmerald itself it looks ok, what I'm not 100% sure is if its something else being used by Esmerald that is becoming irresponsive like the MongoDB access from asyncz (we updated it recently to a newer and better version).

Can you add logging to your application to understand better where it might hang? After how long it becomes irresponsive? I suspect this might be caused by asyncz (which would mean we might need to run an update if you don't have the latest of course or something else that might be causing it to do that).

We can try a couple of things.

1. Add more logging to the who e application to understand where it hangs because like this its hard to understand.
2. Disable temporarily asyncz (which I also think you are actually not using it, just declaring the configuration).

Does this happen in your local when fully testing it? I doubt but still asking.

Out of curiosity, I know Azure webapps with docker, they don't like supervisor so much as per my last experience with it, if you are also using it.

**Reply by tarsil** [reply]
Somehow this is boiling down to my understand of the access to the external resources like redis or the DB and not Esmerald itself.

**Reply by sanket-parte** [reply]
We will add additional logs to debug this.
Currently we are using global redis connection pool, which is a sync client as follows:
![redis_client](https://github.com/user-attachments/assets/e6643625-3aa2-4007-b3be-1e083f8ac013)

**Reply by tarsil** [reply]
Yes I understand but maybe something in the connections is not working so well and can be a myriad of reasons. We have Esmerald applications live running and this is a first but each project is of course, different so the logging is important to try to trace the root cause.

**Comment by Kokoserver** [comment]
Can you try to replicate the application using local redis or mongodb on you dev environment. 

And add logs to check if the redis or mongodb is connecting  then we will know when to do next

**Reply by tarsil** [reply]
@sanket-parte @Kokoserver is also one of the maintainers and he has successfully deployed Esmerald applications for quite a while as well. He is here also to help 👍🏼

**Reply by Kokoserver** [reply]
Hello @sanket-parte

**Reply by sanket-parte** [reply]
Yes, in the development environment, it is working fine since Nginx is not being used there.

**Reply by tarsil** [reply]
One thing I noticed in your nginx that is different from [this config](https://github.com/dymmond/esmerald-tortoise-nginx-gunicorn-supervisor/blob/main/%7B%7B%20cookiecutter.project_name%20%7D%7D/deployment/nginx/nginx.conf) is the proxy pass as the example uses a socket instead of the way you did. Also this is there and I don't see in yours.

```
proxy_connect_timeout 180s;
proxy_read_timeout 300s;

proxy_http_version  1.1;
proxy_cache_bypass  $http_upgrade;
```

And in the supervidor, its not a program, should be `[fcgi-program:gunicorn]` (fcgi) as If i'm not mistaken the program is from sync mode and fcgi will work for async as well

**Reply by sanket-parte** [reply]
Yes, you are right, @tarsil. The example provides socket configuration. However, we are not using a socket to run Gunicorn. Instead, we are binding it to the host (0.0.0.0) and port (8000). This configuration is reflected in the Nginx setup accordingly.

Regarding the supervisord configuration, if I configure it as "fcgi-program," it requires a socket line. That’s why I changed it to "program."

I can try running it with a socket configuration as well.

**Reply by tarsil** [reply]
> Yes, you are right, @tarsil. The example provides socket configuration. However, we are not using a socket to run Gunicorn. Instead, we are binding it to the host (0.0.0.0) and port (8000). This configuration is reflected in the Nginx setup accordingly.
> 
> Regarding the supervisord configuration, if I configure it as "fcgi-program," it requires a socket line. That’s why I changed it to "program."
> 
> I can try running it with a socket configuration as well.

Of course, every project has a different setup and I was not saying you were wrong 👍🏼  just putting out here some differences while we try to understand. Logging will definitely be what will help but if locally works like a charm, so Nginx is causing some issues somehow but again, logging

**Reply by sanket-parte** [reply]
I have also tried a sock-based configuration and tested it without Redis, but the same issue persists. I need to check with Mongoz as well, as I suspect it might be the root cause. There’s nothing left to debug except for this dependency

**Reply by tarsil** [reply]
Are you using Mongoz? Hmm maybe the connection might be causing some issues then. Database connections can be tricky sometimes and usually it's what causes some problems. 

The good thing about Esmerald is that you don't need use dependency injection to use them although you can of course

**Reply by sanket-parte** [reply]
Yes, we are using Mongoz for querying into MongoDB. We also have another Esmerald dummy application deployed without Mongoz, where we are not encountering this issue. Could you tell us how we can debug this Mongoz issue?

**Reply by tarsil** [reply]
Ok, so we established that is not Esmerald which makes it a lot easier.

You mentioned that locally works well so might be some connections between your docker container and MongoDB. 

Can you put logging on Mongoz?

**Reply by sanket-parte** [reply]
Yeah, I will add some logs. Let me provide a few more details: we are using dependency injection to inject the model query manager by applying using() method into our API views. Then, in the API views, we are using those injected model query managers for querying.

**Reply by tarsil** [reply]
Ok it's a matter of design but with Mongoz you can simply use the Model (document) without the need of injecting directly but I guess it's a matter of your design 🙂

**Reply by sanket-parte** [reply]
Yeah we are having multiple mongodb databases, hence we are injecting required database query manager based the request

**Reply by tarsil** [reply]
Makes sense, so might be some of the connections that might not be communicating properly? Hard to know

**Reply by sanket-parte** [reply]
@tarsil, Could you tell us how to enable logs for Mongoz? We are adding a registry object to each model's meta class.

**Reply by tarsil** [reply]
This is a normal python logging object I think, so 

"logger = logging.getLogger("mongoz")"?

Mongoz it's not shipped alwitj unique loggers so that means you can use normal python logging.

**Reply by sanket-parte** [reply]
But for this, do I need to add log statements in the Mongoz source code before running the application?

**Reply by tarsil** [reply]
@sanket-parte no necessarily. You can use a logging library to hook into mongoz library and expose it. You would want to log your queries for example. To see if they are ok or something else

**Reply by sanket-parte** [reply]
@tarsil I have tried the following approach to set up logging, where I obtain a mongoz logger and add a file handler to it. However, the logs are not being written to the file:
```python

class AppSettings(EsmeraldAPISettings)

      def __init__(self, *args: Any, **kwds: Any):
            import logging
            super().__init__(*args, **kwds)
            mongoz_logger = logging.getLogger('mongoz')
            mongoz_logger.setLevel(logging.DEBUG)
            file_handler = logging.FileHandler("/app/logs/mongoz.log")
            formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
            file_handler.setFormatter(formatter)
            file_handler.setLevel(logging.DEBUG)
            mongoz_logger.addHandler(file_handler)

```

The queries are working fine, but when the server encounters a gateway timeout error, the requests are not reaching the server. How can we identify these issues? We need to debug the connection issues. Could you advise where I should place additional logs for better diagnostics?

**Reply by tarsil** [reply]
Ok so when a gateway timeout happens I believe it's on the nginx side. We might need to add logging onto nginx to understand why.

**Reply by sanket-parte** [reply]
I have checked the Nginx logs as well, and it appears that requests were not hitting Nginx at that time

**Reply by tarsil** [reply]
Oh, it's when weirder. How come? If its not even hitting nginx, something else is happening?

**Reply by sanket-parte** [reply]
Hey @tarsil ,
We've tried a different approach: we removed Nginx and configured Gunicorn to use HTTPS, exposing it directly to the frontend. With this setup, the gateway timeout error no longer occurs. It seems that the issue arises when using Nginx as a reverse proxy.

**Reply by tarsil** [reply]
@sanket-parte which is again very odd. @Kokoserver also uses nginx as reverse proxy and no issues occur. Something in the nginx configuration is causing this. 

Btw how do you feel about Esmerald? What do you think of it?

**Reply by sanket-parte** [reply]
We have not stopped the Nginx container; it is still running on the server independently for other applications. We have only removed the Nginx configuration for the Esmerald application.

**Reply by tarsil** [reply]
So is it working like you want now? 

Also, how do you feel about Esmerald? What do you think of it?

**Reply by sanket-parte** [reply]
Yeah, it's a great framework. I've really enjoyed working with Esmerald. Anyone familiar with Django can easily adapt to it. A lot of things are easy to understand, and nothing feels overly complicated. The built-in features are particularly useful in real-world scenarios.

**Reply by tarsil** [reply]
Thank you so much. It was designed to have exactly that feeling so I do really appreciate your input. I'm trying to gather as much testimonials as possible as it would be nice to have them shown since we don't have sponsorship or marketing. Only the good will like yourself. I hope you keep on using it 🙂

**Reply by sanket-parte** [reply]
Yeah, the API server is working fine now. However, when we use Nginx as a reverse proxy, it’s not working as expected. You mentioned that @Kokoserver successfully deployed with Nginx, but we should check a few things—like deploying an Esmerald application that uses MongoDB and mongoz. And configuring with Nginx—since that's our use case.

**Reply by tarsil** [reply]
> Yeah, the API server is working fine now. However, when we use Nginx as a reverse proxy, it’s not working as expected. You mentioned that @Kokoserver successfully deployed with Nginx, but we should check a few things—like deploying an Esmerald application that uses MongoDB and monoz. And configuring with Nginx—since that's our use case.

Actually soon I will be doing that (not date specified) so I can share my inputs with you later on

**Reply by sanket-parte** [reply]
Thank you very much for considering it.

**Reply by sanket-parte** [reply]
Hey @tarsil , did you find time to check this?

**Reply by tarsil** [reply]
@sanket-parte we are trying to replicate this and so far we are having no issues but we will let you know once we have more news.

This looks like more a nginx configuration somehow than anything else. Looks like nginx receives some calls and then dies out. Reason? Well, only if we could debug nginx

**Reply by sanket-parte** [reply]
OK, thanks for the update
